src/vm/analysis/type_checker/natives/mod.rs

```diff
@@ -190,29 +190,44 @@ fn check_special_if(checker: &mut TypeChecker, args: &[SymbolicExpression], cont
 
 fn check_contract_call(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
     check_arguments_at_least(2, args)?;
-    let contract_identifier = match args[0].expr {
-        SymbolicExpressionType::LiteralValue(Value::Principal(PrincipalData::Contract(ref contract_identifier))) => contract_identifier,
-        _ => return Err(CheckError::new(CheckErrors::ContractCallExpectName))
-    };
 
-    let function_name = args[1].match_atom()
+    let func_name = args[1].match_atom()
         .ok_or(CheckError::new(CheckErrors::ContractCallExpectName))?;
     checker.type_map.set_type(&args[1], no_type())?;
 
-    let contract_call_function_type = {
-        if let Some(function_type) = checker.db.get_public_function_type(&contract_identifier, function_name)? {
-            Ok(function_type)
-        } else if let Some(function_type) = checker.db.get_read_only_function_type(&contract_identifier, function_name)? {
-            Ok(function_type)
-        } else {
-            Err(CheckError::new(CheckErrors::NoSuchPublicFunction(contract_identifier.to_string(),
-                                                                  function_name.to_string())))
+    let mut contract_call_args = Vec::new();
+    for arg in args[2..].iter() {
+        if let Some(var_name) = arg.match_trait_reference() {
```

Comment:
Can it ever be the case that some contract-call arguments are trait references, and others are not?